### DISCUSSION TITLE
Questions

### DISCUSSION THREAD

**Discussion by kormax** [kormax/apple-home-key-reader#13] [Q&A]
Questions

**Comment by juaurodri** [comment]
Just a suggestion,  create a buymeacoffee or similar 😊 
Your work on this was top notch

**Reply by kormax** [reply]
Thanks for the suggestion, but I don't want to create even a slightest impression that I've done all of this (and much other) reverse engineering work for profit or any financial interest.

Perhaps in the future, if I ever come around to doing something else, like a full fledged app or a library (been thinking about finally completing my Python NFC library which would support many chips and readers with custom software-based stack), but now now.

**Comment by meistermolli** [comment]
Hi, is there any way to get id run on an esp32? Maybe it's possible to set it up and use it with that? have no Mac or Win nearby our front door.

The reader does work with the esp. but does your software, too?
https://www.youtube.com/watch?v=5Xo8yc4tQYc

Thanks

**Reply by kormax** [reply]
This particular piece of software works only on platforms that are capable of running a full subset of python with OS capabilities - so running it on a MCU as-is is a no go. You can still use an SBC PC like a RPi. That's something the others have done already.

As for ESP32/MCU-compatible solutions, AFAIK there are a couple of projects ongoing, you can look for them either through github search or in my starred repositories.

**Comment by inventor2007** [comment]
Hello, How can I create the card in homekit? Thanks

**Reply by kormax** [reply]
By adding a virtual lock into the Apple Home app.
The pass is created automatically as that happens.

**Reply by inventor2007** [reply]
I use homebridge to add the map and I have the code on [github](https://github.com/inventor2007/nfc-card) but I don’t understand how to put the blocks and uuid I want in the map.

**Reply by kormax** [reply]
I see you're using the initial code made by @KhaosT 

It should be able to cause the card to be added. So that's OK to use.


As for putting "uid" and "blocks" into a "map", I'm not entirely sure what you're referring to.

Going to assume that you've meant that you want to put custom data on a card - **that's impossible**. 
HomeKey is based on public key cryptography, It does not store or transfer any data apart from cryptographic signatures and proofs. And even the keys you don't control, as they're generated at random by device locally.


I think your request comes from assumption that all (physical/virtual) cards act like some data storage.
It is, in fact, true for Mifare - the most common protocol. But in all other cases - the card uses a protocol defined by manufacturer, and most of the other ones don't act as some type of data storage. So you cannot use your HomeKey card at a reader that does not explicitly implement Home Key protocol, and vice versa.

**Reply by inventor2007** [reply]
Ok thanks but if it changes all the time how I can know if it is the right id with my arduino.

**Reply by kormax** [reply]
This can be done by implementing the protocol based on example provided in this exact repository.

**Reply by inventor2007** [reply]
Ok but with an arduino connect to the internet and a rc522 drive I can do it?

**Reply by kormax** [reply]
Yes, in this case It's going to work.

But before you begin, I want you to consider using something like ESP32 instead of a plain Atmel Arduono as it has networking built-in.

Moreover, there are already projects on GitHub that aim at replicating the protocol using those devices, so you don't even need to start from the beginning.

**Reply by inventor2007** [reply]
Do you know a repo with this type of project?

**Reply by kormax** [reply]
> Do you know a repo with this type of project?

[Here's one that looks promising](https://github.com/rednblkx/HomeKey-ESP32).

**Reply by inventor2007** [reply]
And how do you ensure that in this [repo](https://github.com/kormax/apple-enhanced-contactless-polling) you can use the RFID-RC522 module. And can you tell me how you did it.

**Reply by nmanngit** [reply]
Just to make sure, you have to create a virtual lock and cannot associate it directly with an existing one in any way? We've got a Nuki Smart Lock Pro Gen 4 that's connected via Matter. It also supports MQTT.

**Reply by kormax** [reply]
From the POV of homekit this reader is just an another lock, there's so native way of binding as what's done here wasn't supposed to happen.

At this point, we're left with two ways of "binding" them:

1) Via HomeKit Automations

2) Via a web hook call or mqtt message, but you have to add that code by yourself, inside of "on_endpoint_authenticated" method.

**Reply by nmanngit** [reply]
Ok, thanks for the clarification!

**Reply by eigenphase** [reply]
> By adding a virtual lock into the Apple Home app.
> The pass is created automatically as that happens.

I am still confused by this, conceptually. How to add a "virtual lock"? Does it require hardware? If yes, the POC or some existing off-the-shelf lock that supports Apple Home?

When I go "Add Accessory" in the Home app, it wants to scan a QR code or use NFC. I can tap "More options" and then "My Accessory Isn't Shown Here" and then "Enter a Setup Code" but the error is "Unable to connect to accessory", if I enter a random umber.

**Reply by rednblkx** [reply]
Adding a “virtual lock” just means to follow the [steps](https://github.com/kormax/apple-home-key-reader?tab=readme-ov-file#installation--running) of the repository or make your own implementation.

**Reply by eigenphase** [reply]
Thank you! In other words, it does require the hardware already, right? So when creating the key, the phone is already communicating via NFC? 

Other high-level question: Do I need "Apple Home" only for getting the key into Apple Wallet? And afterwards, I can just rely MQTT and even delete the Home app?

**Reply by zerzhang** [reply]
Hi, I'm still a little confused that  how can I add it to my apple Home? Do I need a physical product or can PN532 be added as a device?

NoneType: None
[2024-08-12 15:01:53,239] [    INFO] __init__          :148  searching for reader on path tty:ttyUSB0:pn532
[2024-08-12 15:01:53,239] [    INFO] accessory_driver  :372  Starting accessory NFC Lock on addresses ['192.168.29.128'], port 51926.
To use the QR Code feature, use 'pip install HAP-python[QRCode]'
Enter this code in your HomeKit app on your iOS device: 
[2024-08-12 15:01:59,417] [    INFO] __init__          :151  using PN532v1.6 at /dev/ttyUSB0
[2024-08-12 15:02:20,736] [    INFO] service           :126  Found non-ISODEP Tag with UID: 73BC3929
[2024-08-12 15:02:20,777] [    INFO] service           :130  Waiting for target to leave the field...
[2024-08-12 15:02:30,897] [    INFO] service           :126  Found non-ISODEP Tag with UID: 73BC3929
[2024-08-12 15:02:30,938] [    INFO] service           :130  Waiting for target to leave the field...
[2024-08-12 15:02:31,481] [    INFO] service           :130  Waiting for target to leave the field...
[2024-08-12 15:02:32,027] [    INFO] service           :130  Waiting for target to leave the field...
[2024-08-12 15:02:32,571] [    INFO] service           :130  Waiting for target to leave the field...
[2024-08-12 15:02:33,114] [    INFO] service           :130  Waiting for target to leave the field...
There is no response when the pn532 close to my phone

**Reply by kormax** [reply]
@zerzhang install the Qr code module as suggested in the logs.

After that, upon starting the Application, a QR code will be printed out on the screen. 

Scan it, and you'll be able to connect.

**Reply by zerzhang** [reply]
Hi kormax, Thanks for your help and now I can add it to my home!

**Comment by TBeisch** [comment]
Thank you for this great job!
Can you tell me what I would have to do on the python-side to run two different locks on one system. I would connect two PN532 to two FTDIs which would be connectet to ttyUSB0 and ttyUSB1. What would I do with the program now?

**Reply by kormax** [reply]
Entirely possible. 

You could create a clf instance for each port, and add an extra "for" loop inside of "while" which performs a polling cycle for a clf, thus listening to them both at the same time.

You can also just purchase two PIs, and add them to the same home installation, as Home Key supports any amount of readers and endpoints.

**Reply by TBeisch** [reply]
Thank you for your advice. I‘m afraid I don‘t know enough about Python, so the two PIs will be my solution.

**Reply by kormax** [reply]
Just FYI, I think you can any SBC for this function, not just full Pi3/4/5. 

I've heard that people used Pi/OrangePi Zeroes for running Home Key reader without any issues. So it's not necessary to pay for the fastest and greatest (although you'd get better support with Pis).

**Comment by fuhckos-98** [comment]
I have read over everything but am still a little confused - Once everything is completed, does the PN532 need to continuously be connected to the device (mac/linux) to be utilised, or once everything has been run and set up it can be disconnected for use without being connected to the source?

**Reply by kormax** [reply]
What do you mean by "disconnected" exactly?

Because as far as I understand, you want to disconnect a PN532.

That doesn't really make any sense, as without the NFC reader there is no way to perform NFC authentication, which is the main purpose of this project.

If the only thing you want from this project is a virtual accessory in homekit, you can remove NFC connection code and use it as-is, although you can just create a sample lock accessory based on HAP-python examples on your own.

**Reply by fuhckos-98** [reply]
I may have misunderstood the implementation and I think you have answered my question, but just to clarify that my idea will NOT work as I thought.

My thought process was to use a PC to do everything and have it connected by UART as described. Once this was done I was wanting to unplug the PN532 from the PC and use it as the Home Key Reader without it still being tied to the PC. Essentially so I could have it out the front of my house as a Home Key without the need for wires etc and just have the script running on PC, but I take it it needs the PC in order to verify, unlike a basic NFC tag.

**Reply by kormax** [reply]
PN532 is a "dumb" chip that cannot control itself on its own. It requires a "host", which controls it and implements high-level protocols.

If you don't want to be tied to a PC, you can use a smaller ESP32 based implementation created by @rednblkx.

**Reply by rednblkx** [reply]
Project in question is [here](https://github.com/rednblkx/HomeKey-ESP32)
Keep in mind that it’s not a battery-powered solution so you’ll have to find another way to hook it to power.
